### PR TITLE
Add WeChatComponentAPI and WeChatComponentClient. Ref #50

### PR DESCRIPTION
**Pull request by hunter007**

### CODE REVIEW

**Comment by GitCop** [comment]
There were the following issues with your Pull Request
- Commit: 12b9e0122e66ebba48333dee200cd6ee91ff9a5a
  - Commits must be in the following format: **%{type}(%{scope}): %{description}**
- Commit: c8b5f773c622ffc3a91b8f56930d1a253210c431
  - Commits must be in the following format: **%{type}(%{scope}): %{description}**
- Commit: cf07a468a4b3dd61d7de21296170f1f35e1538cb
  - Commits must be in the following format: **%{type}(%{scope}): %{description}**
- Commit: 8882270f2f9cf3c05149e3df5eb394966990d609
  - Your commit message body contains a line that is longer than 72 characters
  - Commits must be in the following format: **%{type}(%{scope}): %{description}**
- Commit: 89f01141afea3ae0057f74452188339e3fa5c37d
  - Your commit message body contains a line that is longer than 72 characters
  - Commits must be in the following format: **%{type}(%{scope}): %{description}**
- Commit: fd9686d5959ddbac98463b171efdf952b087008f
  - Commits must be in the following format: **%{type}(%{scope}): %{description}**
  - Your subject line is longer than 72 characters

Guidelines are available at https://github.com/jxtech/wechatpy/blob/master/CONTRIBUTING.md

---

This message was auto-generated by https://gitcop.com

File: wechatpy/client/__init__.py [hunk 1]
```diff
@@ -53,3 +54,51 @@ def fetch_access_token(self):
...
+
+        :return: 返回的 JSON 数据包
+        """
+        expires_in = 7200
+        result = self.component.refresh_authorizer_token(
+            self.appid, self.refresh_token)
+        if 'expires_in' in result:
+            expires_in = result['expires_in']
+        self.session.set(
+            'access_token', result['authorizer_access_token'], expires_in)
```
Comment (messense) [inline wechatpy/client/__init__.py]: 这里 session 里存的名字最好加个 component_ 前缀避免使用同一个 session storage 导致和普通 client 重名覆盖？

File: wechatpy/component.py [hunk 2]
```diff
@@ -0,0 +1,351 @@
...
+
+    def _handle_result(self, res, method=None, url=None, **kwargs):
+        result = res.json()
+        if 'errcode' in result:
+            result['errcode'] = int(result['errcode'])
+
+        if 'errcode' in result and result['errcode'] != 0:
+            errcode = result['errcode']
+            errmsg = result['errmsg']
+            if errcode == 42001:
```
Comment (messense) [inline wechatpy/component.py:104]: 确定公众号第三方平台的 errcode 和普通公众号接口 errcode 一致吗？

File: wechatpy/component.py [hunk 3]
```diff
@@ -0,0 +1,351 @@
...
+        return self.session.get('access_token')
+
+    def get(self, url, **kwargs):
+        return self._request(
+            method='get',
+            url_or_endpoint=url,
+            **kwargs
+        )
+
+    _get = get
```
Comment (messense) [inline wechatpy/component.py]: _get 这个可以删掉了，WeChatClient 里面为了保证兼容性遗留的，这里应该不需要。_post 一样。

**Comment by messense** [comment]
Ping #50

File: wechatpy/client/__init__.py (hunk 1 above)
Comment (hunter007) [inline wechatpy/client/__init__.py]: 这是一个问题。因为开放平台下回同时存在多个公众号对象。我再考虑一下

Comment (messense) [inline wechatpy/client/__init__.py]: `'{0}_access_token'.format(self.appid)` 这样应该可以。

File: wechatpy/component.py (hunk 2 above)
Comment (hunter007) [inline wechatpy/component.py:104]: 官方没有说明。这个我在最近几天会再次测试

File: wechatpy/component.py (hunk 3 above)
Comment (hunter007) [inline wechatpy/component.py]: OK

File: wechatpy/client/__init__.py (hunk 1 above)
Comment (hunter007) [inline wechatpy/client/__init__.py]: 是的。另外，现在的 session 机制支持 django 吗？因为我们的项目是基于 django 和 celery 的。我们希望通过celery来刷新 access_token。celery 内部是直接使用 django 的 cache 来更新access_token。

Comment (hunter007) [inline wechatpy/client/__init__.py]: 建议发布一个新版本， alpha 的。等这块测试完成，在发布正式版本。

Comment (messense) [inline wechatpy/client/__init__.py]: 没有支持，但是可以直接继承 ``wechatpy.session.SessionStorage` 实现 `get`, `set` 和 `delete` 方法自定义 session 存储方式，这样应该就可以了。

Comment (messense) [inline wechatpy/client/__init__.py]: 稍微改一下，合并了之后发布一个 1.1-alpha 版本吧

Comment (hunter007) [inline wechatpy/client/__init__.py]: 简单化是可以的。但 django 这边提供可选参数，毕竟不是完整支持。这个需求我再考虑怎么做。能现发布 alpha 版吗？我先用着，同时也对代码做测试。

Comment (messense) [inline wechatpy/client/__init__.py]: 麻烦先稍微改一下吧，这里和删掉 _get/_post 就好了，errcode 先不管

**Comment by GitCop** [comment]
There were the following issues with your Pull Request
- Commit: 12b9e0122e66ebba48333dee200cd6ee91ff9a5a
  - Commits must be in the following format: **%{type}(%{scope}): %{description}**
- Commit: c8b5f773c622ffc3a91b8f56930d1a253210c431
  - Commits must be in the following format: **%{type}(%{scope}): %{description}**
- Commit: cf07a468a4b3dd61d7de21296170f1f35e1538cb
  - Commits must be in the following format: **%{type}(%{scope}): %{description}**
- Commit: 8882270f2f9cf3c05149e3df5eb394966990d609
  - Your commit message body contains a line that is longer than 72 characters
  - Commits must be in the following format: **%{type}(%{scope}): %{description}**
- Commit: 89f01141afea3ae0057f74452188339e3fa5c37d
  - Your commit message body contains a line that is longer than 72 characters
  - Commits must be in the following format: **%{type}(%{scope}): %{description}**
- Commit: fd9686d5959ddbac98463b171efdf952b087008f
  - Commits must be in the following format: **%{type}(%{scope}): %{description}**
  - Your subject line is longer than 72 characters
- Commit: 1322e750912ceeb49fe5b392dd40c7402c006ea7
  - Commits must be in the following format: **%{type}(%{scope}): %{description}**

Guidelines are available at https://github.com/jxtech/wechatpy/blob/master/CONTRIBUTING.md

---

This message was auto-generated by https://gitcop.com

**Comment by hunter007** [comment]
在吗？已经更新了。

**Comment by messense** [comment]
pip install wechatpy==1.1a0.1

**Comment by hunter007** [comment]
Big Thanks!

# 

赵文涛  steven zhao

Tel：（+86） 186 5168 3851

# Email：wentao79@gmail.com

2015-09-14 20:43 GMT+08:00 messense notifications@github.com:

**Comment by hunter007** [comment]
​我在本地跑测试的时候一直出错，怎么回事儿知道吗？我是 mac 环境。

/wechatpy/.tox/py27/lib/python2.7/site-packages/nose/importer.py", line 94,
in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/Users/steven/Projects/wechatpy/tests/test_enterprise_parser.py",
line 5, in <module>
    from wechatpy.enterprise import parse_message
  File "/Users/steven/Projects/wechatpy/wechatpy/enterprise/**init**.py",
line 6, in <module>
    from wechatpy.enterprise.crypto import WeChatCrypto  # NOQA
  File "/Users/steven/Projects/wechatpy/wechatpy/enterprise/crypto.py",
line 3, in <module>
    from wechatpy.crypto import BasePrpCrypto, BaseWeChatCrypto
  File "/Users/steven/Projects/wechatpy/wechatpy/crypto/**init**.py", line
22, in <module>
    from wechatpy.crypto.base import BasePrpCrypto
  File "/Users/steven/Projects/wechatpy/wechatpy/crypto/base.py", line 16,
in <module>
    raise Exception('You must install either cryptography or PyCrypto!')
Exception: You must install either cryptography or PyCrypto!

刚刚打包的1.1版本出异常了。现在我要修复它。所以要炮筒所有用例。

​

# 

赵文涛  steven zhao

Tel：（+86） 186 5168 3851

# Email：wentao79@gmail.com

2015-09-14 20:45 GMT+08:00 赵文涛 wentao79@gmail.com:

**Comment by messense** [comment]
pip install cryptography

或者

pip install pycrypto

**Comment by hunter007** [comment]
我在系统环境和env内都安装过了，不行。
于此相关的6个盈用例一直跑不过
2015年9月14日 下午9:55，"messense" notifications@github.com写道：

**Comment by messense** [comment]
那你可以试试用 tox 跑，

pip install tox

然后

tox -e py27-pycrypto 或者 tox -e py27-cryptography

**Comment by hunter007** [comment]
OK。我试试。
2015年9月14日 下午10:04，"messense" notifications@github.com写道：

File: wechatpy/component.py (hunk 2 above)
Comment (hunter007) [inline wechatpy/component.py:104]: 确定错误码是一致的：https://open.weixin.qq.com/cgi-bin/showdocument?action=dir_list&t=resource/res_list&verify=1&id=open1419318634&token=ad56f6cae48d5732165898a9ae4cd0a59a47eff2&lang=zh_CN

Comment (messense) [inline wechatpy/component.py:104]: OK. Thanks.

**Pull request merged**